rfd/0036/user-stories/jupiter-example-com-jenkins.md

```diff
@@ -0,0 +1,77 @@
+
+### Jenkins jobs
+
+Jenkins should be configured with two jobs, one for development and one for production. The separate production job allows for manual approval of the build while reducing risk of accidental deployment (see below).
+
+### Jenkins development/test job
+
+This Jenkins job is configured to receive GitHub webhook pushes. The Jenkins job should run build and deployment tasks defined in the code repository so that the process for deployment is owned by the team that owns the code.
+
+Each GitHub branch has an associated Mariposa project. The Jenkins job clones a template manifest (from the repo) for each new branch. Each changeset received on a given branch will be deployed as a new version of that manifest.
+
+The job task might look like the following:
+
+- Run `docker-compose build` locally
+- Tags containers with the `GIT_BRANCH` and `GIT_COMMIT`
+- Containers to a registry (optionally a private registry)
+- Render manifest template to manifest
+- `mariposa push manifest`
+- `mariposa deploy version`
+
+Mariposa will then see a new version of the manifest and deploy those containers for testing.
+
+> Open question: when do you shut down containers or delete a project associated with a branch? We can receive `delete` events from GitHub web hooks, but it’s not clear Jenkins jobs can do anything with them.
+
+
+### Jenkins production job
+
+Despite the automation for continuous build and deploy of the staging projects, the deploys to the live site require manual promotion. After reviewing the relevant staging site and any logs generated by Jenkins, the operators merge a PR to master, triggering a new build of master that is tagged with the date and time of the build (the builds collect resources beyond the scope of the git repo, so the git SHA is not a consistent identifier). The master branch is built and staged using the same logic as the development jobs above in a project named <repo-name>-branch-master.
+
+After testing and previewing that final staging build, the operators can choose to manually deploy that build to the live site. This manual deployment is done via a separate production job. This Jenkins job is not configured to receive GitHub webhook pushes. Instead, it is a parameterized job which deploys the parameterized tag (with a default like “production”). The end-user tags the release in git with a version and the tag.
+
+```
+git tag v1.2.3
+git tag production
+git push --tags
+```
+
+The operator uses the Jenkins UI (or CLI) to start the deployment job with the particular tag needed (with the default being part of the Jenkins job definition). The job tags the correct version as “production” and renders the production manifest with the new version and pushes it to Mariposa for deployment. Note that Mariposa isn’t watching registry server; you need to explicitly update images in IMGAPI to get new images, so updating the Docker image tag in the registry is not enough to trigger a deployment.
```

Comment:
Agreed that tagging suitably addresses (1).

Can we address (2) by incorporating the git SHA into the build? Similar to what we do in [ContainerPilot](https://github.com/joyent/containerpilot/blob/master/makefile#L10) or [Joyent.com](https://github.com/joyent/joyent.com-kirby/pull/181/files#diff-d0b2ef35f4d7cb0284dd5cf566b62f51R307). This makes the git SHA indelibly marked in the image layers.

Resolving (3) by making that same git SHA show up in the project manifest is a tiny bit harder because we probably can't enforce any of this workflow (it's all owned by the end user's Jenkins job). But we could easily expose the `Labels` from the Docker images associated with a manifest version though. Maybe we provide a conventional label to the end users? (like `triton.mariposa.version` or something)